server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierTimeTracker.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.service;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.metrics.CounterMetric;
+import org.elasticsearch.common.xcontent.ToXContentFragment;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.core.TimeValue;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+public final class ClusterApplierTimeTracker {
+
+    // Most listeners are added when the node is initialized, however a few
+    // are added during the lifetime of a node and for these cases we need ConcurrentHashMap
+    private final Map<String, CounterMetric> timeSpentPerApplier = new ConcurrentHashMap<>();
+    private final Map<String, CounterMetric> timeSpentPerListener = new ConcurrentHashMap<>();
+
+    void addApplier(String name) {
+        timeSpentPerApplier.put(name, new CounterMetric());
+    }
+
+    void removeApplier(String name) {
+        timeSpentPerApplier.remove(name);
+    }
+
+    void addListener(String name) {
+        timeSpentPerListener.put(name, new CounterMetric());
+    }
+
+    void removeListener(String name) {
+        timeSpentPerListener.remove(name);
+    }
+
+    void incrementTimeSpentApplier(String name, TimeValue took) {
+        increment(timeSpentPerApplier.get(name), took);
+    }
+
+    void incrementTimeSpentListener(String name, TimeValue took) {
+        increment(timeSpentPerListener.get(name), took);
+    }
+
+    private static void increment(CounterMetric counter, TimeValue took) {
+//        assert counter != null : "no counter for [" + name + "]";
+        if (counter != null) {
+            // Listeners/appliers may get removed while executing.
+            counter.inc(took.millis());
+        }
+    }
+
+    Stats getStats() {
+        return new Stats(
```

Comment:
I would prefer if the stats we returned were from a completed update so they're all consistent with each other. No need for anything fancy tho, the updates are all on the applier thread and we don't get stats often so we can just do something simple with `synchronised` I think.

That said, it's a bit more complicated to return consistent stats if we let other threads add and remove the trackers. Could we instead add trackers as needed during cluster state application, and remove any ones that didn't execute at the end of a successful update?

Reply:
>  No need for anything fancy tho, the updates are all on the applier thread and we don't get stats often so we can just do something simple with synchronised I think.

👍 

>  Could we instead add trackers as needed during cluster state application, and remove any ones that didn't execute at the end of a successful update?

👍